src/main/java/org/wildfly/security/sasl/util/AuthenticationTimeoutSaslServerFactory.java

```diff
@@ -0,0 +1,146 @@
+ * chain of SASL server factories.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+public final class AuthenticationTimeoutSaslServerFactory extends AbstractDelegatingSaslServerFactory {
+
+    /**
+     * The default amount of time, in seconds, after which an authentication attempt should time out.
+     */
+    public static final long DEFAULT_TIMEOUT = 300;
+
+    private final ScheduledExecutorService scheduledExecutorService;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param delegate the delegate {@code SaslServerFactory}
+     * @param scheduledExecutorService the scheduled executor to use to handle authentication timeout tasks
+     */
+    public AuthenticationTimeoutSaslServerFactory(final SaslServerFactory delegate, final ScheduledExecutorService scheduledExecutorService) {
+        super(delegate);
+        Assert.checkNotNullParam("scheduledExecutorService", scheduledExecutorService);
+        this.scheduledExecutorService = scheduledExecutorService;
+    }
+
+    @Override
+    public SaslServer createSaslServer(final String mechanism, final String protocol, final String serverName, final Map<String, ?> props, final CallbackHandler cbh) throws SaslException {
+        final long timeout;
+        if (props.containsKey(WildFlySasl.AUTHENTICATION_TIMEOUT)) {
+            timeout = Long.parseLong((String) props.get(WildFlySasl.AUTHENTICATION_TIMEOUT));
+        } else {
+            timeout = DEFAULT_TIMEOUT;
+        }
+        final NameAssignedCallbackHandler nameAssignedCallbackHandler = new NameAssignedCallbackHandler(cbh);
+        final SaslServer delegateSaslServer = delegate.createSaslServer(mechanism, protocol, serverName, props, nameAssignedCallbackHandler);
+        return delegateSaslServer == null ? null : new AbstractDelegatingSaslServer(delegateSaslServer) {
+            private final AtomicBoolean complete = new AtomicBoolean();
+            private final AtomicBoolean nameAssigned = new AtomicBoolean();
+            private volatile ScheduledFuture<Void> timeoutTask;
+
+            @Override
+            public byte[] evaluateResponse(final byte[] response) throws SaslException {
+                try {
+                    final byte[] challenge = delegate.evaluateResponse(response);
+                    if (nameAssignedCallbackHandler.isNameAssigned() && nameAssigned.compareAndSet(false, true)) {
+                        // Schedule a task to terminate the authentication attempt if it takes too long
+                        timeoutTask = scheduledExecutorService.schedule(() -> {
+                            dispose();
+                            return null;
+                        }, timeout, TimeUnit.SECONDS);
+                    }
+                    if (isComplete() && complete.compareAndSet(false, true)) {
+                        cancelTimeoutTask();
```

Comment:
It might be a good idea to do this check before we schedule & execute the timeout task.